homeassistant/components/zwave_js/sensor.py

```diff
@@ -47,6 +84,124 @@
 LOGGER = logging.getLogger(__name__)
 
 
+@dataclass
+class ZwaveSensorBaseEntityDescription(SensorEntityDescription):
+    """Base description of a ZwaveSensorBase entity."""
+
+    info: ZwaveDiscoveryInfo | None = None
```

Comment:
An alternative to passing the info to the entity description is that we use the info to calculate a composed id first and then use that to look up the matching entity description. The id can be composed of the different types of groups we have, scale type, sensor type etc. The id will always have the same structure, ie number of groups, but it will sometimes have None for a group when that group isn't important for the description. The id will be turned into a string and compared with the entity description `key` attribute when looking up the entity description.

The entity descriptions can be statically defined with known id permutations based on the existing cases, in a constant container. The container would contain about 20 different entity descriptions based on the current number of cases.

Downside with this approach is that we have to write more entity descriptions but the upside is that the entity description definitions are simpler.

Reply:
I have mixed feelings about this suggestion. The other downside with this approach (IMO at least) is that we somewhat move sensor platform logic into the discovery module. But I suppose another way of looking at it is that we are discovering additional information about the value to provide the platform in order to configure it. So I don't know.

I will give this approach a try later and see how it looks. The main purpose of pushing this initial version out was to validate that the constants I created upstream would provide utility for this refactor so definitely open to iterating on the design.

Reply:
Do you think it would make sense to reuse data templates for this approach? We create data templates for each of the CC's that then get resolved to a platform specific data structure which can be used to populate the entity descriptions. We could even move all of the logic in there such that if a data template is available to the callback to add entities, it just passes each value from the resolved data to the entity description constructor.